### PR TITLE
Add jsx-restrict-classnames rule

### PR DESCRIPTION
**Pull request by niki4810**
This rule prevents react components from using a pre-defined set of restricted classnames, or classnames which start with a certain prefix. 

***Use case scenarios:***
- Deprecating old class names
- Prevent developers from using common class such as `grid`, `btn` and enforce them to use components like `<Buttons/>`, `<Grid/>.`
- Restrict prefix class names, for e.g. `js-*` that are often copy pasted code.

***screencap:***
eslint-demo-2

//cc @yannickcr @baer

### CODE REVIEW

**Comment by niki4810** [comment]
@yannickcr Ping for review :)

File: package.json [hunk 1]
```diff
@@ -23,7 +23,7 @@
   "homepage": "https://github.com/yannickcr/eslint-plugin-react",
   "bugs": "https://github.com/yannickcr/eslint-plugin-react/issues",
   "devDependencies": {
-    "babel-eslint": "5.0.0-beta6",
+    "babel-eslint": "5.0.0-beta9",
```
Comment (baer) [inline package.json]: Are you sure you meant to commit this?

Comment (niki4810) [inline package.json]: @baer yes that was intentional. When running test via `npm test` i hit this issue.

```
➜  eslint-plugin-react git:(master) ✗ npm run test

> eslint-plugin-react@3.16.1 test /Users/nkatak1/devr/eslint-plugin-react
> npm run lint && npm run unit-test


> eslint-plugin-react@3.16.1 lint /Users/nkatak1/devr/eslint-plugin-react
> eslint ./


> eslint-plugin-react@3.16.1 unit-test /Users/nkatak1/devr/eslint-plugin-react
> istanbul cover --dir reports/coverage node_modules/mocha/bin/_mocha tests/**/*.js -- --reporter dot



  ․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․TypeError: Cannot read property 'visitClass' of undefined
    at monkeypatch (/Users/nkatak1/devr/eslint-plugin-react/node_modules/babel-eslint/index.js:220:40)
    at Object.exports.parse (/Users/nkatak1/devr/eslint-plugin-react/node_modules/babel-eslint/index.js:362:5)
    at parse (/Users/nkatak1/devr/eslint-plugin-react/node_modules/eslint/lib/eslint.js:574:27)
    at EventEmitter.module.exports.api.verify (/Users/nkatak1/devr/eslint-plugin-react/node_modules/eslint/lib/eslint.js:701:19)
    at runRuleForItem (/Users/nkatak1/devr/eslint-plugin-react/node_modules/eslint/lib/testers/rule-tester.js:315:38)
    at testValidTemplate (/Users/nkatak1/devr/eslint-plugin-react/node_modules/eslint/lib/testers/rule-tester.js:333:26)
    at Context.<anonymous> (/Users/nkatak1/devr/eslint-plugin-react/node_modules/eslint/lib/testers/rule-tester.js:421:25)
    at callFn (/Users/nkatak1/devr/eslint-plugin-react/node_modules/mocha/lib/runnable.js:286:21)
    at Test.Runnable.run (/Users/nkatak1/devr/eslint-plugin-react/node_modules/mocha/lib/runnable.js:279:7)
    at Runner.runTest (/Users/nkatak1/devr/eslint-plugin-react/node_modules/mocha/lib/runner.js:421:10)
    at /Users/nkatak1/devr/eslint-plugin-react/node_modules/mocha/lib/runner.js:528:12
    at next (/Users/nkatak1/devr/eslint-plugin-react/node_modules/mocha/lib/runner.js:341:14)
    at /Users/nkatak1/devr/eslint-plugin-react/node_modules/mocha/lib/runner.js:351:7
    at next (/Users/nkatak1/devr/eslint-plugin-react/node_modules/mocha/lib/runner.js:283:14)
    at Immediate._onImmediate (/Users/nkatak1/devr/eslint-plugin-react/node_modules/mocha/lib/runner.js:319:5)
    at processImmediate [as _immediateCallback] (timers.js:383:17)
```

upon lookup I found that updating eslint-babel dependency fixed the issue. So updated that dependency.

Comment (niki4810) [inline package.json]: just saw in the latest master, this peer dependency has been updated, will revert this change.

**Comment by niki4810** [comment]
@yannickcr could you please review it when you get a chance. Let me know if anything needs to be improved.

**Comment by yannickcr** [comment]
Thanks for this. But, even if I'm ok with the purpose of the rule, I'm a bit reluctant to merge this rule in its actual form since it has a very different way to work beside other rules.

- Other rules does not offer a way to customize the error message, so it is a bit weird to have this functionality for only this rule.
- Usage of RegExp is imho more powerful and flexible than a `isPrefix` property (see the `sort-comp` rule for example).

Also it does not support the official way to manage conditional classes in React components ([classnames](https://github.com/jedwatson/classnames)) but this can be a future enhancement.

**Comment by niki4810** [comment]
Thanks for the feedback @yannickcr. I agree, this rule is not very generic so wanted to get your feedback. Closing this PR.
